core-bundle/src/Twig/Interop/PhpTemplateProxyNode.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Interop;
+
+use Twig\Compiler;
+use Twig\Node\Node;
+
+class PhpTemplateProxyNode extends Node
+{
+    public function __construct(string $extensionName)
+    {
+        parent::__construct([], [
+            'extension_name' => $extensionName,
+        ]);
+    }
+
+    public function compile(Compiler $compiler): void
+    {
+        /*
+         * echo $this->extensions["Contao\\…\\ContaoExtension"]->renderLegacyTemplate(
+         *     $this->getTemplateName(),
+         *     array_map(
+         *         static function(callable $block) use ($context): string {
+         *             ob_start(); $block($context); return ob_get_clean();
+         *         }, $blocks
+         *     ), $context
+         * );
+         */
+
+        $compiler
+            ->write('echo $this->extensions[')
+            ->repr($this->getAttribute('extension_name'))
+            ->raw(']->renderLegacyTemplate('."\n")
+            ->indent()
+            ->write('$this->getTemplateName(),'."\n")
+            ->write('array_map('."\n")
+            ->indent()
+            ->write('static function(callable $block) use ($context): string {'."\n")
+            ->indent()
+            ->write('ob_start(); $block($context); return ob_get_clean();'."\n")
```

Comment:
Can/should we wrap the closing of the output buffer in a `finally {}`?

```suggestion
            ->write('ob_start(); try { $block($context); return ob_get_contents(); } finally { ob_end_clean(); }'."\n")
```

Reply:
Twig does it like this for its own "renderBlock" (see \Twig\Template):

```php
if ($this->env->isDebug()) {
    ob_start();
} else {
    ob_start(function () { return ''; });
}

// render the thing…

return ob_get_clean();
```

Maybe we should go this route as well? I'm not sure I fully understand the consequences of each version. :slightly_smiling_face:

Reply:
I think we should do both:

```php
if ($this->env->isDebug()) {
    ob_start();
} else {
    ob_start(function () { return ''; });
}

try { 
    $block($context);
    return ob_get_contents(); 
} finally {
    ob_end_clean();
}
```

As far as I understand the code: The empty handler function ensures that the output never gets accidentally output to the browser, even in cases like memory limit, timeout, segfault. The `finally` ensures that the opened output buffer gets closed even if the render code throws an unhandled exception.

Reply:
implemented in  e63889a63c48c0fac6253ed0d6bcedf81f704e84